### DISCUSSION TITLE
Any idea how to render Latex equations nicely in Telegram?

### DISCUSSION THREAD

**Discussion by 10bn** [n3d1117/chatgpt-telegram-bot#494] [General]
I'm using the bot mostly for studying, do you guys have a suggestion hot render latex equations nicely in Telegram?

**Comment by chaserCN** [comment]
I've created a plugin that converts LaTeX to SVG using https://math.vercel.app/ and then converts the SVG to PNG in the code. You can check it out here: https://github.com/chaserCN/chatgpt-telegram-bot/tree/custom

**Reply by alexk-epsilon** [reply]
@chaserCN your plugin is great 👌

I did \[x^2 + 2*x + 3 =0 \] and it rendered:
![Screen Shot 2024-01-18 at 4 40 16 PM](https://github.com/n3d1117/chatgpt-telegram-bot/assets/150628561/f4467126-fb7a-4e8d-9923-13bc875636b9)

Smashing 👍

Then, I asked it to solve the equation:

> solve \[x^2 + 2*x + 3 =0 \]

and it gave me a long brain dump in raw latex:
```
The equation \(x^2 + 2x + 3 = 0\) does not have real solutions since the discriminant \((b^2 - 4ac)\) is negative. However, we can find the complex solutions using the quadratic formula. The quadratic formula states that for an equation of the form \(ax^2 + bx + c = 0\), the solutions for \(x\) can be calculated as:

[x = \frac{-b \pm \sqrt{b^2 - 4ac}}{2a}\]

In our case, \(a = 1\), \(b = 2\), and \(c = 3\). Substituting these values into the quadratic formula, we have:

[x = \frac{-2 \pm \sqrt{2^2 - 4(1)(3)}}{2(1)}\]

Simplifying further, we get:

[x = \frac{-2 \pm \sqrt{-8}}{2}\]
```

Since we have a square root of a negative number, we can rewrite it using the imaginary unit \(i\), where \(i^2 = -1\):

[x = \frac{-2 \pm \sqrt{8}i}{2}\]

Simplifying the expression, we have:

[x = -1 \pm \sqrt{2}i\]

Therefore, the solutions to the equation \(x^2 + 2x + 3 = 0\) are: 
[x = -1 + \sqrt{2}i\]
[x = -1 - \sqrt{2}i\]

I wonder if there is a way to apply plugin to the model output as well...

**Reply by chaserCN** [reply]
I usually ask it "show the answer as an image" and it understands. 
![2024-01-19 02 56 55](https://github.com/n3d1117/chatgpt-telegram-bot/assets/5011333/b20771c3-1a17-4e9d-9ae1-bc2662d0dd2d)

You can try "solve [x^2 + 2*x + 3 =0 ] and show the solution as an image"

**Reply by alexk-epsilon** [reply]
Yeah. 
![Screen Shot 2024-01-19 at 9 59 42 AM](https://github.com/n3d1117/chatgpt-telegram-bot/assets/150628561/fc4e39cb-d0f9-4254-99ce-326184cb4fe3)
Not sure why it doesn't run latex() plugin on individual formulas in response to `solve [x^2 + 2*x + 3 =0 ]` though . It should be fairly obvious to the model that they're in latex, or are we missing something...

**Reply by alexk-epsilon** [reply]
Hi folks. I looked at various online tex-to-image converters but none seemed to be robust enough or easy enough to use to handle multi-line LaTex  in a reasonable way. It is also worth noting that the model rarely outputs proper latex (but I firmly believe that it will improve eventually lol) So, I decided to capitalize on @chaserCN 's awesome idea  to actually use local LaTex converter via sympy.preview() 
The source  is available at  https://github.com/alexk-epsilon/chatgpt-telegram-bot/tree/local_latex
Here are some results:
![telegram-cloud-photo-size-1-5010504196520259175-x](https://github.com/n3d1117/chatgpt-telegram-bot/assets/150628561/ddddb4f1-ef54-4dd9-9b42-e246efdd5634)
@chaserCN thank you for the inspiration and please let me know if you want me to contribute this to your repo.
